File: 906-walking-robot-simulation/walking-robot-simulation.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        obsMap = dict()
        for x,y in obstacles:
            if x not in obsMap:
                obsMap[x] = set()
            obsMap[x].add(y)
        x,y,theta = 0,0,-90
        direction = [(1,0),(0,1),(-1,0),(0,-1)]
        currDirection = 1
        res = 0
        for c in commands:
            if c==-1:
                # +4 to map the negative value to corresponding direction
                currDirection = ((currDirection - 1)+4)%4
                print(c,"rotate +90",currDirection)
            elif c==-2:
                currDirection = (currDirection + 1) % 4
                print(c,"rotate -90",currDirection)
            else:
                dx,dy = direction[currDirection]
                print(f"dx={dx}, dy={dy}")
                for i in range(1,c+1):
                    newX = x+dx if dx != 0 else x
                    newY = y+dy if dy != 0 else y
                    
                    print(c,f"next: i={i}",newX, newY)
                    if newY not in obsMap.get(newX,set()):
                        x,y = newX, newY
                        res = max(res, x**2 + y**2)
                    else:
                        print("obstacle ahead")
                        break
        return res

File: 906-walking-robot-simulation/walking-robot-simulation.java
class Solution {
    public int robotSim(int[] commands, int[][] obstacles) {
        Map<Integer,Set<Integer>> obsMap = new HashMap<>();
        for(int[] p: obstacles){
            obsMap.putIfAbsent(p[0],new HashSet<>());
            obsMap.get(p[0]).add(p[1]);
        }

        int x=0, y=0, cd=1;
        int[][] directions = new int[][]{
            {1,0},
            {0,1},
            {-1,0},
            {0,-1}
        };
        int res = 0;
        for (int c:commands){
            if (c==-1){
                cd = ((cd - 1)+4)%4;
            }else if (c==-2){
                cd = (cd + 1)%4;
            }else{
                int[] d = directions[cd];
                for (int i=1;i<c+1;i++){
                    int newX = d[0]==0?x:(x + d[0]);
                    int newY = d[1]==0?y:(y + d[1]);
                    if (obsMap.containsKey(newX) && obsMap.get(newX).contains(newY)){
                        break;
                    }else{
                        x=newX;
                        y=newY;
                        res = Math.max(res,x*x + y*y);
                    }
                }
            }
        }
        return res;
    }
}